From: =?utf-8?b?Ik1hcmNvIFRyZXZpc2FuIChUcmV2acOxbyki?= <mail@3v1n0.net>
Date: Thu, 7 Dec 2017 01:33:01 +0100
Subject: workspaceThumbnail: Disconnect from window signals on destruction

Avoid to try to destroy the window clone content more than once
when a window is destroyed, and do it in proper order.

Bug: https://bugzilla.gnome.org/show_bug.cgi?id=791233
Bug: https://gitlab.gnome.org/GNOME/gnome-shell/issues/1
Forwarded: https://gitlab.gnome.org/GNOME/gnome-shell/merge_requests/4
(cherry picked from commit 956fc4c11c6473b23a1011421f0786c65c882de1)
---
 js/ui/workspaceThumbnail.js | 13 +++++++------
 1 file changed, 7 insertions(+), 6 deletions(-)

diff --git a/js/ui/workspaceThumbnail.js b/js/ui/workspaceThumbnail.js
index f5f6913..f33c1b0 100644
--- a/js/ui/workspaceThumbnail.js
+++ b/js/ui/workspaceThumbnail.js
@@ -70,12 +70,7 @@ var WindowClone = new Lang.Class({
 
         this.clone._updateId = this.metaWindow.connect('position-changed',
                                                        Lang.bind(this, this._onPositionChanged));
-        this.clone._destroyId = this.realWindow.connect('destroy', Lang.bind(this, function() {
-            // First destroy the clone and then destroy everything
-            // This will ensure that we never see it in the _disconnectSignals loop
-            this.clone.destroy();
-            this.destroy();
-        }));
+        this.clone._destroyId = this.realWindow.connect('destroy', this.destroy.bind(this));
         this._onPositionChanged();
 
         this.actor.connect('button-release-event',
@@ -142,6 +137,12 @@ var WindowClone = new Lang.Class({
     },
 
     destroy: function () {
+        // First destroy the clone and then destroy everything
+        // This will ensure that we never see it in the _disconnectSignals loop
+        this.metaWindow.disconnect(this.clone._updateId);
+        this.realWindow.disconnect(this.clone._destroyId);
+        this.clone.destroy();
+
         this.actor.destroy();
     },
 
